README
Antoni Stevenet
2557398
itzmjauz@gmail.com


Changelog
Changelog for the code handout belonging to the 
Systems Programming Course 2015
Vrije Universiteit Amsterdam (VU-FEW-CompSci)
------

1.5
---
Updated aud_readinit() description
Updated copyright notices.

1.4
---
Author contact info : wdb -_at-_ few.vu.nl
Lost in history.

1.3
---

Cleaned up the mess
	
1.2
----

exchanged incorrect swap_long for swap_short in aud_readinit (wh->bit...)
exchanged incorrect swap_short for swap_long in aud_readinit (wh->rate..)

----
1.1

merged Erik van der Kouwe's patches:
	fix Solaris issues (nonexistent string.h and stdint.h)
	fix Cygwin issues (compilation flags)

----
1.0

initial release



audio.c
/* audio.[ch]
 *
 * part of the Systems Programming assignment
 * (c) Vrije Universiteit Amsterdam, 2005-2015. BSD License applies
 * author  : wdb -_at-_ few.vu.nl
 * contact : arno@cs.vu.nl
 * */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define AUDIODEV	"/dev/dsp"
#define swap_short(x)		(x)
#define swap_long(x)		(x)

#include <stdint.h>
#include <sys/soundcard.h>
#include "audio.h"

/* Definitions for the WAVE format */

#define RIFF		"RIFF"	
#define WAVEFMT		"WAVEfmt"
#define DATA		0x61746164
#define PCM_CODE	1

typedef struct _waveheader {
  char		main_chunk[4];	/* 'RIFF' */
  uint32_t	length;		/* filelen */
  char		chunk_type[7];	/* 'WAVEfmt' */
  uint32_t	sc_len;		/* length of sub_chunk, =16 */
  uint16_t	format;		/* should be 1 for PCM-code */
  uint16_t	chans;		/* 1 Mono, 2 Stereo */
  uint32_t	sample_fq;	/* frequence of sample */
  uint32_t	byte_p_sec;
  uint16_t	byte_p_spl;	/* samplesize; 1 or 2 bytes */
  uint16_t	bit_p_spl;	/* 8, 12 or 16 bit */ 

  uint32_t	data_chunk;	/* 'data' */
  uint32_t	data_length;	/* samplecount */
} WaveHeader;

int aud_readinit (char *filename, int *sample_rate, 
		  int *sample_size, int *channels ) 
{
  /* Sets up a descriptor to read from a wave (RIFF). 
   * Returns file descriptor if successful*/
  int fd;
  WaveHeader wh;
		  
  if (0 > (fd = open (filename, O_RDONLY))){
    fprintf(stderr,"unable to open the audiofile\n");
    return -1;
  }

  read (fd, &wh, sizeof(wh));

  if (0 != bcmp(wh.main_chunk, RIFF, sizeof(wh.main_chunk)) || 
      0 != bcmp(wh.chunk_type, WAVEFMT, sizeof(wh.chunk_type)) ) {
    fprintf (stderr, "not a WAVE-file\n");
    errno = 3;// EFTYPE;
    return -1;
  }
  if (swap_short(wh.format) != PCM_CODE) {
    fprintf (stderr, "can't play non PCM WAVE-files\n");
    errno = 5;//EFTYPE;
    return -1;
  }
  if (swap_short(wh.chans) > 2) {
    fprintf (stderr, "can't play WAVE-files with %d tracks\n", wh.chans);
    return -1;
  }
	
  *sample_rate = (unsigned int) swap_long(wh.sample_fq);
  *sample_size = (unsigned int) swap_short(wh.bit_p_spl);
  *channels = (unsigned int) swap_short(wh.chans);	
	
  fprintf (stderr, "%s chan=%u, freq=%u bitrate=%u format=%hu\n", 
	   filename, *channels, *sample_rate, *sample_size, wh.format);
  return fd;
}

int aud_writeinit (int sample_rate, int sample_size, int channels) 
{
  /* Sets up the audio device params. 
   * Returns device file descriptor if successful*/
  int audio_fd, error;
  char *devicename;

  printf("requested chans=%d, sample rate=%d sample size=%d\n", 
	 channels, sample_rate, sample_size);
	
  if (NULL == (devicename = getenv("AUDIODEV")))
    devicename = AUDIODEV;
    	
  if ((audio_fd = open (devicename, O_WRONLY, 0)) < 0) {
    perror ("setparams : open ") ;
    return -1;
  } 
  if (ioctl (audio_fd, SNDCTL_DSP_RESET, 0) != 0) {
    perror ("setparams : reset ") ;
    close(audio_fd);
    return -1;	
  } 
	
  if ((error = ioctl (audio_fd, SNDCTL_DSP_SAMPLESIZE, &sample_size)) != 0) {
    perror ("setparams : bitwidth ") ;
    close(audio_fd);
    return -1;
  } 

  if (ioctl (audio_fd, SNDCTL_DSP_CHANNELS, &channels) != 0) {
    perror ("setparams : channels ") ;
    close(audio_fd);
    return -1;
  } 

  if ((error = ioctl (audio_fd, SNDCTL_DSP_SPEED, &sample_rate)) != 0) {
    perror ("setparams : sample rate ") ;
    close(audio_fd);
    return -1;
  } 

  if ((error = ioctl (audio_fd, SNDCTL_DSP_SYNC, 0)) != 0) {
    perror ("setparams : sync ") ;
    close(audio_fd);
    return -1;
  } 
  printf("set chans=%d, sample rate=%d sample size=%d\n",
	 channels,sample_rate, sample_size);
  return audio_fd;
}





audio.h
/* audio.[ch]
 *
 * part of the Systems Programming assignment
 * (c) Vrije Universiteit Amsterdam, 2005-2015. BSD License applies
 * author  : wdb -_at-_ few.vu.nl
 * contact : arno@cs.vu.nl
 * */

/** open a WAV-file for reading
 *
 * this function checks whether the file pointed to by filename exists,
 * and if it is of the right type. Note that not all WAV filetypes are
 * supported. Only the simplest uncompressed PCM streams can be read.
 *
 * the function writes metadata about the opened file in the integers pointed
 * to by the parameters.
 *
 * @param filename	a non-NULL pointer to a string denoting a local file
 * @param sample_rate	the number of samples per second, e.g., 44kHz
 * @param sample_size	the precision of the wave-approximations, e.g., 16bit
 * @param channels	the number of channels. No THX or dobly, we only accept mono and stereo
 *
 * @return a new file descriptor on success, <0 on failure
 */
int aud_readinit (char *filename, int *sample_rate, int*sample_size, int* channels );

/** write an uncompressed PCM stream to the speaker
 *
 * this function opens the speaker device as a regular file. You can write to the returned
 * file descriptor as if it was a normal file.
 *
 * @params sample_rate, sample_size, channels: see aud_readinit above
 * @return a new descriptor or <0 if an error occurred. NB: 0 itself is a valid descriptor number!
 */
int aud_writeinit (int sample_rate, int sample_size, int channels); 



client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <netdb.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/select.h>
#include <sys/time.h>
#include <arpa/inet.h>

#include "audio.h"

#define BUFSIZE 1024
#define PORT 32581
#define DONESIZE 4
#define MAX_TIMEOUT 4

static int breakloop = 0; 
const char * DENY = "DENY";
const char * CONNECT = "CONNECT";
const char * FERROR = "FERROR";
const char * VOL = "VOL";
const char * NOISE = "NOISE";
const char * SPEED = "SPEED";

int recvtimeout(int fd) {
  struct timeval timeout;
  fd_set read_set;
  int nb;

  FD_ZERO(&read_set);
  FD_SET(fd, &read_set);
  timeout.tv_sec = MAX_TIMEOUT;
  timeout.tv_usec = 0;

  nb = select(fd + 1, &read_set, NULL, NULL, &timeout);
  if (FD_ISSET(fd, &read_set)) {
    return 1;
  }
  else {
    return nb;
  }
}

void sigint_handler(int sigint) {
  if (breakloop) {  
    printf("SIGINT occurred, shutting down...\n");
    exit(-1);
  }
  breakloop = 1;
  printf("SIGINT caught, closing client.., force quit with ctrl-c\n");
}

struct in_addr resolvehost(char * hostname) {
  struct hostent *resolv;
  struct in_addr *addr;

  resolv = gethostbyname(hostname);
  if (resolv == NULL) {
    //adress not found
    exit(1);
  }
  addr = (struct in_addr*) resolv->h_addr_list[0];
  
  return *addr;
}
   
int comms(int server_fd, int audio_fd, struct sockaddr_in addr) { //COMMUNICATIONS
  int bytesread,  err;
  char buf[BUFSIZE], okmssg[] = "OK";
  socklen_t flen = sizeof(struct sockaddr_in);

  bytesread = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
  while (bytesread > 0) {
    write(audio_fd, buf, bytesread);
    err = sendto(server_fd, okmssg, strlen(okmssg), 0, (struct sockaddr *) &addr, flen);
    if (err < 0) {
      printf("Cannot send OK to server\n");
      return -1;
    }
    err = recvtimeout(server_fd);
    if (err < 0) {
      printf("ERROR : Connection error\n");
      return -1;
    } else if (err == 0) {
      printf("ERROR : Connection lost\n");
      return -1;
    } 
    bytesread = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
    
    if(strncmp(buf, "DONE", DONESIZE) == 0) {
      bytesread = -1;
    }
  }
  printf("Got final packet!\n");

  return 1;
}

int initconnection(int server_fd, char * hostname, char * filename, int filter, char * filterarg) {
  struct sockaddr_in sock;
  int err = 0, audio_fd, sample_rate, sample_size, channels;
  struct in_addr addr;
  char buf[BUFSIZE] = "";
  socklen_t flen = sizeof(struct sockaddr_in);
  addr = resolvehost(hostname);

  sock.sin_family         = AF_INET;
  sock.sin_port           = htons(PORT);
  sock.sin_addr           = addr;

  if((sizeof(filename) + strlen(CONNECT)) < BUFSIZE) {
    strncat(buf, CONNECT, strlen(CONNECT)); //connect message
    strncat(buf, filename, strlen(filename));
  } else {
    printf("Entered filename too long! enter something smaller than %d\n", BUFSIZE);
    return -1;
  }

  err = sendto(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &sock, flen);
  if (err < 0) {
    printf("Cannot send packets to server!, quitting\n");
    perror("sendto");
    return -1;
  }

  printf("Waiting for answer from server...\n");
  err = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &sock, &flen);
  if (err < 0) return -1;
  if (strncmp(buf, DENY, strlen(DENY)) == 0) {
    printf("ERROR, Server denied connection, probably busy\n");
    return -1;
  }
  if (strncmp(buf + strlen(CONNECT), filename, strlen(filename)) == 0) {
    printf("Connection OK!\n");
    err = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &sock, &flen);
    if (err < 0) {
      printf("ERROR : cannot receive audio info from server, exiting..\n");
      return -1;
    }
    if (strncmp(buf, FERROR, strlen(FERROR)) == 0) {
      printf("ERROR : Audiofile not found! Terminating!\n");
      return -1;
    }

    sample_size = strtol(strtok(buf, "|"), NULL, 10);
    sample_rate = strtol(strtok(NULL, "|"), NULL, 10);
    channels = strtol(strtok(NULL, "|"), NULL, 10);
    
    if (filter != 3) {
      audio_fd = aud_writeinit(sample_rate, sample_size, channels);//open audio output
    } else {
      audio_fd = aud_writeinit(sample_rate * strtol(filterarg, NULL, 10) / 100, sample_size, channels);//open audio output
      printf("%d\n", sample_rate * 45 / 100);
    }
    if (audio_fd < 0) {
      printf("ERROR : Cannot open audio output\n");
      return -1;
    }
    
    memset(buf, 0, BUFSIZE);    
    if(filter != 0 && (strlen(filterarg) + 1) < BUFSIZE) {
      buf[0] = (char) filter + '0';
      strncat(buf, filterarg, strlen(filterarg));
      strcat(buf, "\0");
    } else {
      buf[0] = '0';
    }

    err = sendto(server_fd, buf, strlen(buf), 0, (struct sockaddr *) &sock, flen);
    if (err < 0) {
      printf("ERROR : Could not send filter data, terminating\n");
      return -1;
    }
    
    comms(server_fd, audio_fd, sock);

  } else {
    printf("Got wrong server response, quitting..\n");
    return -1;
  }

  return -1;
}

int main(int argc, char ** argv) {
  int server_fd, filter = 0;

  printf("Audio client, by Antoni stevenet @ 2016\n"); //logging so that user knows stuff is happening

  signal(SIGINT, sigint_handler);  //catch ctrl-c

  if (argc < 3 || argc > 5) { //check parameters
    printf("ERROR : Incorrect amount of parameters passed\n"); 
    printf("Usage : %s <hostname/IP> <filename> [<filter> [filter options]]\n", argv[0]);
    return -1;
  }

  server_fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP); //server connection
  if (server_fd < 0) {
    printf("ERROR : cannot open input connection");
    return -1;
  }

  if(argv[3]) {
    if(strncmp(argv[3], VOL, strlen(VOL) + 1) == 0) { 
      filter = 1;
      if(!argv[4]) {
        printf("ERROR : Not enough arguments for VOL filter, add percentage of volume to play as argument\n");
        return -1;
      }
      if(strlen(argv[4]) > 3) {
        printf("ERROR : filter argument too big! should be smaller than 1000\n");
        return -1;
      }
    } else if(strncmp(argv[3], NOISE, strlen(NOISE) + 1) == 0) { 
      filter = 2;
      if(!argv[4]) {
        printf("ERROR : Not enough arguments for NOISE filter, add number of HZ to add as sine wave as argument\n");
        return -1;
      }
      if(strlen(argv[4]) > 5) {
        printf("ERROR filter argument too big! should be smaller than 10000 \n");
        return -1;
      }
    } else if(strncmp(argv[3], SPEED, strlen(SPEED) + 1) == 0) { 
      filter = 3;
      if(!argv[4]) {
        printf("ERROR : Not enough arguments for SPEED filter, add percentage to multiply rate by\n");
        return -1;
      }
      if(strlen(argv[4]) > 3) {
        printf("ERROR filter argument too big! should be smaller than 1000\n");
        return -1;
        if(argv[4][0] == '0') {
          printf("SPEED argument can't be 0 or start with 0, retry..\n");
          return -1;
        }
      }
    } else {
      printf("ERROR filter not found! try VOL NOISE or SPEED filter instead\n");
      return -1;
    }
  }

  initconnection(server_fd, argv[1], argv[2], filter, argv[4]);

  return 0;
}


audioclient.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <netdb.h>
#include <sys/types.h>
#include <sys/socket.h>

#include "library.h"
#include "audio.h"

#define BUFSIZE 1024
#define PORT 32581
#define DONESIZE 4

static int breakloop = 0; 

void sigint_handler(int sigint) {
  if (breakloop) {  
    printf("SIGINT occurred, shutting down...\n");
    exit(-1);
  }
  breakloop = 1;
  printf("SIGINT caught, closing client.., force quit with ctrl-c\n");
}

struct in_addr resolvehost(char * hostname) {
  struct hostent *resolv;
  struct in_addr *addr;

  resolv = gethostbyname(hostname);
  if (resolv == NULL) {
    //adress not found
    exit(1);
  }
  addr = (struct in_addr*) resolv->h_addr_list[0];

  return *addr;
}
   
int comms(int server_fd, int audio_fd, struct sockaddr_in addr) { //COMMUNICATIONS
  int bytesread, count = 0;
  char buf[BUFSIZE], okmssg[] = "OK";
  socklen_t flen = sizeof(struct sockaddr_in);

  bytesread = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
  while (bytesread > 0) {
    write(audio_fd, buf, bytesread);
    sendto(server_fd, okmssg, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
    count++;
    printf("Got packet: %d\n", count);
    bytesread = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
    
    if(strncmp(buf, "DONE", DONESIZE) == 0) {
      bytesread = -1;
    }
  }
  printf("Got final packet!\n");

  return 1;
}

int initconnection(int server_fd, char * hostname, char * filename) {
  struct sockaddr_in sock;
  int err = 0, audio_fd, sample_rate, sample_size, channels;
  struct in_addr addr;
  char buf[BUFSIZE];
  socklen_t flen = sizeof(struct sockaddr_in);
  addr = resolvehost(hostname);

  sock.sin_family         = AF_INET;
  sock.sin_port           = htons(PORT);
  sock.sin_addr           = addr;

  err = sendto(server_fd, filename, BUFSIZE, 0, (struct sockaddr *) &sock, flen);
  if (err < 0) {
    printf("Cannot send packets to server!, quitting\n");
    return -1;
  }


  printf("Waiting for answer from server...\n");
  err = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &sock, &flen);
  if (err < 0) return -1;

  if (strncmp(buf, filename, strlen(filename)) == 0) {
    printf("Connection OK!\n");
    err = recvfrom(server_fd, buf, BUFSIZE, 0, (struct sockaddr *) &sock, &flen);
    if (err < 0) {
      printf("ERROR : cannot receive audio info from server, exiting..\n");
      return -1;
    }

    sample_size = strtol(strtok(buf, "|"), NULL, 10);
    sample_rate = strtol(strtok(NULL, "|"), NULL, 10);
    channels = strtol(strtok(NULL, "|"), NULL, 10);

    audio_fd = aud_writeinit(sample_rate, sample_size, channels);//open audio output
    if (audio_fd < 0) {
      printf("ERROR : Cannot open audio output\n");
      return -1;
    }
    comms(server_fd, audio_fd, sock);

  } else {
    printf("Got wrong server response, quitting..\n");
    return -1;
  }

  return -1;
}

int main(int argc, char ** argv) {
  int server_fd;

  printf("Audio client, by Antoni stevenet @ 2016\n"); //logging so that user knows stuff is happening

  signal(SIGINT, sigint_handler);  //catch ctrl-c

  if (argc < 3) { //check parameters
    printf("ERROR : Incorrect amount of parameters passed\n"); 
    printf("Usage : %s <hostname/IP> <filename>\n", argv[0]);
    return -1;
  }

  server_fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP); //server connection
  if (server_fd < 0) {
    printf("ERROR : cannot open input connection");
    return -1;
  }
 
  initconnection(server_fd, argv[1], argv[2]);
  //TODO LATER , library stuff

  return 0;
}


audioserver.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <arpa/inet.h>
#include <sys/time.h>
#include <sys/select.h>

#include "library.h"
#include "audio.h"

#define BUFSIZE 1024
#define PORT 32581

static int breakloop = 0;

int recvtimeout(int fd) {
  struct timeval timeout;
  fd_set read_set;
  int nb;

  FD_ZERO(&read_set);
  FD_SET(fd, &read_set);
  timeout.tv_sec = 1;
  timeout.tv_usec = 0;

  nb = select(fd + 1, &read_set, NULL, NULL, &timeout);
  if (FD_ISSET(fd, &read_set)) {
    return 1;
  }
  else {
    return nb;
  }
}

int streamfile(int fd, int data_fd, struct sockaddr_in addr, float sleeptime) {
  int bytesread, err = 0, count = 0;
  socklen_t flen = sizeof(struct sockaddr_in);
  char buf[BUFSIZE], okmssg[] = "OK";

  bytesread = read(data_fd, buf, BUFSIZE);
  while (bytesread > 0) {
    usleep(sleeptime);
    err = sendto(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
    if (err < 0) {
      printf("ERROR : cannot send audio packet, skipping request\n");
      return -1;
    }
    printf("Sending packet ...");   // not doing this to remove delay in audioplay
    err = recvtimeout(fd);
    if (err < 0) {
      printf("ERROR : Connection error\n");
      return -1;
    } else if (err == 0) {
      printf("Packet : lost\n");
    } else {
      err = recvfrom(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
      if (err < 0) {
        printf("ERROR : Connection error\n");
      }
      printf(" packet %d correctly received!\n", count);
    }
    if(strncmp(buf, okmssg, 2) != 0) {
      printf("\nNo OK received, skipping request\n");
      return -1;
    }
    count++;

    bytesread = read(data_fd, buf, BUFSIZE);
  }

  printf("Sending DONE packet to client...\n");
  strncpy(buf, "DONE", 4);
  err = sendto(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, flen); 
  
  if (err < 0) {
    printf("ERROR : failed sending DONE packet, continue anyway..\n");
    return -1;
  } else {
    printf("Succcesfuly streamed audiofile, waiting for new connections..\n");
    return -1;
  }
}

int stream_audio(int fd, socklen_t flen, char * filename, struct sockaddr_in addr) {
  int data_fd, channels, sample_size, sample_rate, bitrate, err;
  float sleeptime;
  char buf[BUFSIZE];

  data_fd = aud_readinit(filename, &sample_rate, &sample_size, &channels);
  if (data_fd < 0) {
    printf("ERROR : failed to open audio file : %s\n, Skipping request\n", filename);
    return -1;
  }

  printf("Audiofile found! sending details...\n");
  sprintf(buf, "%d|%d|%d|", sample_size, sample_rate, channels);

  err = sendto(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
  if (err < 0) {
    printf("Could not send packet...Skipping request\n");
    return -1;
  }
   
  bitrate = sample_size * sample_rate * channels;
  sleeptime = (1000000 * (float) (BUFSIZE * 8)) / (float) bitrate; //time per packet in microseconds

  streamfile(fd, data_fd, addr, sleeptime);
  return 1;
}

void connect_client(int fd, socklen_t flen) {
  int err = 0;
  char buf[BUFSIZE];
  struct sockaddr_in addr;
  
  err = recvfrom(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
  if (err < 0) {
    printf("ERROR : %d, when attempting initial connection\n", err);
    connect_client(fd, flen);
  } else {
    printf("Connected to host : %s, port : %d: %s\n", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port), buf);
    err = sendto(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
    
    if (err < 0) {
      printf("Cannot send messages to host, skipping client\n");
      connect_client(fd, flen);
      return;
    }

    stream_audio(fd, flen, buf, addr);
    connect_client(fd, flen);
    return;
  }
}

void sigint_handler(int sigint) {
  if (breakloop) {
    printf("SIGINT occurred, shutting down...\n");
    exit(-1);
  }
  breakloop = 1;
  printf("SIGINT caught, closing client.., force quit with ctrl-c\n");
}

int main(int argc, char ** argv) {
  struct sockaddr_in addr;
  int fd, err = 0;
  socklen_t flen;

  printf("Audio server, by Antoni Stevenet @ 2016\n");

  signal(SIGINT, sigint_handler); //catch ctrl-c

  fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP); //server connection
  if (fd < 0) {
    printf("ERROR : cannot open input connection\n");
    return -1;
  }

  addr.sin_family         = AF_INET;
  addr.sin_port           = htons(PORT);
  addr.sin_addr.s_addr    = htonl(INADDR_ANY);

  printf("Binding socket..\n");
  flen = sizeof(struct sockaddr_in);
  err = bind(fd, (struct sockaddr *) &addr, flen);
  if (err < 0) {
    printf("ERROR : cannot bind socket\n");
    close(fd);
    return -1;
  }

  printf("Starting communication loop\n");
  while (!breakloop) {
    //wait for connection
    //after connection stream 
    connect_client(fd, flen);
    sleep(1);
  }

  return 0;
}



server.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <arpa/inet.h>
#include <sys/time.h>
#include <sys/select.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <stdint.h>
#include <math.h>
#include <limits.h>

#include "audio.h"

#define BUFSIZE 1024
#define PORT 32581
#define MAX_PACKET_LOSS 3

static int breakloop = 0;
static int filter = 0;
static int filterarg = 0;
static int sample_size = 0;
static int sample_rate = 0;
char * DONE = "DONE";
char * CONNECT = "CONNECT";
char * FERROR = "FERROR";
char * DENY = "DENY";
char * OK = "OK";

int recvtimeout(int fd) {
  struct timeval timeout;
  fd_set read_set;
  int nb;

  FD_ZERO(&read_set);
  FD_SET(fd, &read_set);
  timeout.tv_sec = 1;
  timeout.tv_usec = 0;

  nb = select(fd + 1, &read_set, NULL, NULL, &timeout);
  if (FD_ISSET(fd, &read_set)) {
    return 1;
  }
  else {
    return nb;
  }
}

float getPercent(int source) {
  float percentage;
  percentage = (float) source / 100; 
  return percentage; 
}

char * noise(char * buffer) {
  if(sample_size == 16) {
    for(int i = 0 ; i < BUFSIZE ; i++) {
      buffer[i] += (int16_t) (SHRT_MAX * 0.1f * sinf(2.0f * M_PI * filterarg * i / sample_rate));
    }
  } else if(sample_size == 8) {
    for(int i = 0 ; i < BUFSIZE ; i++) {
      buffer[i] += (int8_t) (SHRT_MAX * 0.1f * sinf(2.0f * M_PI * filterarg * i / sample_rate));
    }
  } else {
    printf("ERROR : Filter does not support given sample size\n");
    exit(-1);
  }

  return buffer;
}

char * vol(char * buffer) {
  int sampleCount;
  float percent = getPercent(filterarg);
  
  if(sample_size == 16) {
    int16_t * samp = (int16_t *) &buffer[0];
    sampleCount = (BUFSIZE / sizeof(int16_t));
    for (int i = 0; i < sampleCount; i++) {
      samp[i] = (int16_t) (samp[i] * percent);
    }
  } else {
    int8_t * samp = (int8_t *) &buffer[0];
    sampleCount = (BUFSIZE / sizeof(int8_t));
    for (int i = 0; i < sampleCount; i++) {
      samp[i] = (int8_t) (samp[i] * percent);
    }
  }
   
  return buffer;
}

float adjustsleep(float sleeptime) {
  return sleeptime / getPercent(filterarg);
}

char * filterfunc(char * buffer) {
  if(filter == 1) return vol(buffer);
  if(filter == 2) return noise(buffer);
  else return buffer;
}

int streamfile(int fd, int data_fd, struct sockaddr_in addr, float sleeptime) {
  int bytesread, err = 0, lostcount = 0;
  socklen_t flen = sizeof(struct sockaddr_in);
  char buf[BUFSIZE], * filtered;

  if(filter == 3) sleeptime = adjustsleep(sleeptime); 
  bytesread = read(data_fd, buf, BUFSIZE);
  while (bytesread > 0) {
    filtered = filterfunc(buf);
    usleep(sleeptime);
    err = sendto(fd, filtered, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
    if (err < 0) {
      printf("ERROR : cannot send audio packet, skipping request\n");
      return -1;
    }
    err = recvtimeout(fd);
    if (err < 0) {
      printf("ERROR : Connection error\n");
      return -1;
    } else if (err == 0) {
      printf("Packet : lost\n");
      lostcount++;
      if(lostcount > MAX_PACKET_LOSS) {
        printf("Connection lost, skipping request\n");
        return -1;
      }
    } else {
      err = recvfrom(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
      if (err < 0) {
        printf("ERROR : Connection error\n");
      }
    }
    if(strncmp(buf, OK, strlen(OK)) != 0) {
      if(strncmp(buf, CONNECT, strlen(CONNECT)) == 0) {
        printf("Sending DENY to new client\n");
        err = sendto(fd, DENY, strlen(DENY), 0, (struct sockaddr *) &addr, flen);
        if(err < 0) {
          printf("ERROR, Connection error\n");
        }
      } else {
        if(lostcount > MAX_PACKET_LOSS) {
          printf("Connection lost, skipping request\n");
          return -1;
        }
        printf("No OK received, retrying\n");
        lostcount++;
      }
    } else {
      lostcount = 0;
    }

    if(lostcount == 0) bytesread = read(data_fd, buf, BUFSIZE);
  }

  printf("Sending DONE packet to client...\n");
  err = sendto(fd, DONE, strlen(DONE), 0, (struct sockaddr *) &addr, flen); 
  
  if (err < 0) {
    printf("ERROR : failed sending DONE packet, continue anyway..\n");
    return -1;
  } else {
    printf("Succcesfuly streamed audiofile, waiting for new connections..\n");
    return -1;
  }
}

int stream_audio(int fd, socklen_t flen, char * filename, struct sockaddr_in addr) {
  int data_fd, channels, bitrate, err;
  float sleeptime;
  char buf[BUFSIZE];

  data_fd = aud_readinit(filename + strlen(CONNECT), &sample_rate, &sample_size, &channels);
  if (data_fd < 0) {
    printf("ERROR : failed to open audio file : %s\n, Skipping request\n", filename + strlen(CONNECT));
    err = sendto(fd, FERROR, strlen(FERROR), 0, (struct sockaddr *) &addr, flen);
    if (err < 0) {
      printf("ERROR : Couldn't send File error message\n");
      return -1;
    }
    return -1;
  }

  printf("Audiofile found! sending details...\n");
  sprintf(buf, "%d|%d|%d|", sample_size, sample_rate, channels);

  err = sendto(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
  if (err < 0) {
    printf("Could not send packet...Skipping request\n");
    return -1;
  }
  //read filter stuff
  memset(buf, 0, BUFSIZE);
  err = recvfrom(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
  filter = buf[0] - '0';
  filterarg = strtol(buf + 1, NULL, 10);
   
  bitrate = sample_size * sample_rate * channels;
  sleeptime = (1000000 * (float) (BUFSIZE * 8)) / (float) bitrate; //time per packet in microseconds
  streamfile(fd, data_fd, addr, sleeptime);
  return 1;
}

void connect_client(int fd, socklen_t flen) {
  int err = 0;
  char buf[BUFSIZE];
  struct sockaddr_in addr;
  
  err = recvfrom(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, &flen);
  if (err < 0) {
    printf("ERROR : %d, when attempting initial connection\n", err);
    connect_client(fd, flen);
  } else if (strncmp(buf, CONNECT, strlen(CONNECT)) != 0) {
    printf("Not a connect message!, skipping!\n");
    connect_client(fd, flen);
    return;
  } else {
    printf("Connected to host : %s, port : %d\n", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
    err = sendto(fd, buf, BUFSIZE, 0, (struct sockaddr *) &addr, flen);
    
    if (err < 0) {
      printf("Cannot send messages to host, skipping client\n");
      connect_client(fd, flen);
      return;
    }

    stream_audio(fd, flen, buf, addr);
    connect_client(fd, flen);
    return;
  }
}

void sigint_handler(int sigint) {
  if (breakloop) {
    printf("SIGINT occurred, shutting down...\n");
    exit(-1);
  }
  breakloop = 1;
  printf("SIGINT caught, closing client.., force quit with ctrl-c\n");
}

int main(int argc, char ** argv) {
  struct sockaddr_in addr;
  int fd, err = 0;
  socklen_t flen;

  printf("Audio server, by Antoni Stevenet @ 2016\n");

  signal(SIGINT, sigint_handler); //catch ctrl-c

  fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP); //server connection
  if (fd < 0) {
    printf("ERROR : cannot open input connection\n");
    return -1;
  }

  addr.sin_family         = AF_INET;
  addr.sin_port           = htons(PORT);
  addr.sin_addr.s_addr    = htonl(INADDR_ANY);

  printf("Binding socket..\n");
  flen = sizeof(struct sockaddr_in);
  err = bind(fd, (struct sockaddr *) &addr, flen);
  if (err < 0) {
    printf("ERROR : cannot bind socket\n");
    close(fd);
    return -1;
  }

  printf("Starting communication loop\n");
  while (!breakloop) {
    //wait for connection
    //after connection stream 
    connect_client(fd, flen);
    sleep(1);
  }

  return 0;
}

